### PR TITLE
Lock factory to allow using umock in multithreaded tests

### PR DESCRIPTION
**Pull request by dcristoloveanu**

### CODE REVIEW

File: devdoc/umock_lock_factory_windows.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+```
+
+## static lock functions
+
+```c
+static void umock_lock_windows_acquire_shared(UMOCK_C_LOCK_HANDLE lock);
+static void umock_lock_windows_release_shared(UMOCK_C_LOCK_HANDLE lock);
+static void umock_lock_windows_acquire_exclusive(UMOCK_C_LOCK_HANDLE lock);
+static void umock_lock_windows_release_exclusive(UMOCK_C_LOCK_HANDLE lock);
+static void umock_lock_windows_release_destroy(UMOCK_C_LOCK_HANDLE lock);
```
Comment (mattdurak) [inline devdoc/umock_lock_factory_windows.md:20]: >_release [](start = 30, length = 8)

probably just destroy without release

File: devdoc/umock_lock_factory_pthread.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+```
+
+## static lock functions
+
+```c
+static void umock_lock_pthread_acquire_shared(UMOCK_C_LOCK_HANDLE lock);
+static void umock_lock_pthread_release_shared(UMOCK_C_LOCK_HANDLE lock);
+static void umock_lock_pthread_acquire_exclusive(UMOCK_C_LOCK_HANDLE lock);
+static void umock_lock_pthread_release_exclusive(UMOCK_C_LOCK_HANDLE lock);
+static void umock_lock_pthread_release_destroy(UMOCK_C_LOCK_HANDLE lock);
```
Comment (mattdurak) [inline devdoc/umock_lock_factory_pthread.md:20]: >release [](start = 31, length = 7)

destroy

File: doc/umock_c.md [hunk 3]
```diff
@@ -1328,3 +1329,54 @@ If `ENABLE_MOCKS` is not defined, `IMPLEMENT_MOCKABLE_FUNCTION` shall expand to
...
 
 `UMOCK_REAL` shall produce the name of the real function generated by umock.
+
+### umock_c lock factory
+
+```c
+int umock_c_init_with_lock_factory(ON_UMOCK_C_ERROR on_umock_c_error, UMOCK_C_LOCK_FACTORY_CREATE_LOCK_FUNC lock_factory_create_lock, void* lock_factory_create_lock_params);
+```
+
+There are scenarios where mocking is done in multi threaded environments.
```
Comment (mattdurak) [inline doc/umock_c.md:1339]: >mocking is done [](start = 26, length = 15)

"mocked calls are made" (sounds a bit more clear to me in terms of what this solves, multi-threaded calls to mock functions)

File: src/umock_c.c [hunk 4]
```diff
@@ -49,20 +53,22 @@ int umock_c_init(ON_UMOCK_C_ERROR on_umock_c_error)
         /* Codes_SRS_UMOCK_C_LIB_01_152: [ const void\* ]*/
         /* Codes_SRS_UMOCK_C_01_023: [ umock_c_init shall initialize the umock types by calling umocktypes_init. ]*/
         if ((umocktypes_init() != 0) ||
-            /* Codes_SRS_UMOCK_C_01_002: [ umock_c_init shall register the C naive types by calling umocktypes_c_register_types. ]*/
```
Comment (mattdurak) [inline src/umock_c.c:52]: >naive [](start = 77, length = 5)

lol

File: src/umock_lock_factory_pthread.c [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+static void umock_lock_pthreads_acquire_shared(UMOCK_C_LOCK_HANDLE lock)
+{
+    if (lock == NULL)
+    {
+        /* Codes_SRS_UMOCK_LOCK_FACTORY_PTHREAD_01_010: [ If lock is NULL, umock_lock_pthread_acquire_shared shall return. ]*/
+        UMOCK_LOG("Invalid arguments: UMOCK_C_LOCK_HANDLE lock=%p", lock);
+    }
+    else
+    {
+        UMOCK_C_LOCK_PTHREADS* umock_c_lock_windows = (UMOCK_C_LOCK_PTHREADS*)lock;
```
Comment (mattdurak) [inline src/umock_lock_factory_pthread.c:29]: >windows [](start = 44, length = 7)

not windows here

File: src/umockcallrecorder.c [hunk 6]
```diff
@@ -23,25 +28,122 @@ typedef struct UMOCKCALLRECORDER_TAG
...
+
+static void internal_lock_release_shared_if_needed(UMOCKCALLRECORDER_HANDLE call_recorder)
+{
+    if (call_recorder->lock != NULL)
+    {
+        call_recorder->lock->release_shared(call_recorder->lock);
+    }
+}
+
+void internal_umockcallrecorder_reset_all_calls(UMOCKCALLRECORDER_HANDLE umock_call_recorder)
```
Comment (mattdurak) [inline src/umockcallrecorder.c:68]: >internal_umockcallrecorder_reset_all_calls [](start = 5, length = 42)

static

File: src/umockcallrecorder.c [hunk 7]
```diff
@@ -269,82 +377,89 @@ const char* umockcallrecorder_get_expected_calls(UMOCKCALLRECORDER_HANDLE umock_
...
                             break;
                         }
                         else
                         {
-                            umock_call_recorder->expected_calls_string = new_expected_calls_string;
-                            (void)memcpy(umock_call_recorder->expected_calls_string + current_length, stringified_call, stringified_call_length + 1);
-                            current_length += stringified_call_length;
-                        }
+                            size_t stringified_call_length = strlen(stringified_call);
+                            new_expected_calls_string = umockalloc_realloc(umock_call_recorder->expected_calls_string, current_length + stringified_call_length + 1);
```
Comment (mattdurak) [inline src/umockcallrecorder.c:409]: >new_expected_calls_string = umockalloc_realloc [](start = 28, length = 46)

smells like exclusive locking should be used...

File: src/umockcallrecorder.c [hunk 8]
```diff
@@ -366,68 +481,75 @@ const char* umockcallrecorder_get_actual_calls(UMOCKCALLRECORDER_HANDLE umock_ca
...
+        internal_lock_acquire_shared_if_needed(umock_call_recorder);
         {
-            new_actual_calls_string = (char*)umockalloc_realloc(umock_call_recorder->actual_calls_string, 1);
-            if (new_actual_calls_string == NULL)
+            if (umock_call_recorder->actual_call_count == 0)
             {
-                /* Codes_SRS_UMOCKCALLRECORDER_01_026: [ If allocating memory for the resulting string fails, umockcallrecorder_get_actual_calls shall fail and return NULL. ]*/
-                UMOCK_LOG("umockcallrecorder: Cannot allocate memory for actual calls.");
-                result = NULL;
+                new_actual_calls_string = umockalloc_realloc(umock_call_recorder->actual_calls_string, 1);
```
Comment (mattdurak) [inline src/umockcallrecorder.c:489]: >mockalloc_realloc(umock_call_recorder->actual_calls_string [](start = 43, length = 58)

Also seems like exclusive would be needed here

File: src/umockcallrecorder.c [hunk 9]
```diff
@@ -475,107 +604,140 @@ UMOCKCALLRECORDER_HANDLE umockcallrecorder_clone(UMOCKCALLRECORDER_HANDLE umock_
...
-        if (result != NULL)
+        /* Codes_SRS_UMOCKCALLRECORDER_01_082: [ If a lock was created for the call recorder umock_call_recorder, umockcallrecorder_clone shall acquire_the lock in shared mode for umock_call_recorder. ]*/
+        internal_lock_acquire_shared_if_needed(umock_call_recorder);
         {
-            size_t i;
-
-            result->expected_calls = (UMOCK_EXPECTED_CALL*)umockalloc_malloc(sizeof(UMOCK_EXPECTED_CALL) * umock_call_recorder->expected_call_count);
-            if (result->expected_calls == NULL)
+            /* Codes_SRS_UMOCKCALLRECORDER_01_037: [ If allocating memory for the new umock call recorder instance fails, umockcallrecorder_clone shall fail and return NULL. ]*/
+            result = umockcallrecorder_create(NULL, NULL);
```
Comment (mattdurak) [inline src/umockcallrecorder.c:611]: >NULL, NUL [](start = 46, length = 9)

why not just pass the factory in here? Shouldn't that save some lines below?

File: tests/umockcallrecorder_ut/CMakeLists.txt [hunk 10]
```diff
@@ -11,6 +11,7 @@ umockcallrecorder_ut.c
 
 set(${theseTestsName}_c_files
 umockcallrecorder_mocked.c
+../../src/
```
Comment (mattdurak) [inline tests/umockcallrecorder_ut/CMakeLists.txt:14]: does this add all of the source files in that path? Or just a mistake (line should be removed)?

File: tests/umock_c_with_lock_factory_int/minipal/umock_threadapi_pthread.c [hunk 11]
```diff
@@ -0,0 +1,127 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#define _DEFAULT_SOURCE
```
Comment (mattdurak) [inline tests/umock_c_with_lock_factory_int/minipal/umock_threadapi_pthread.c:4]: >_DEFAULT_SOURCE [](start = 8, length = 15)

what's this?

File: tests/umock_c_with_lock_factory_int/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+else()
+# For others assume pthreads (this should be a real check for pthreads)
+set(test_platform_c_files
+./minipal/umock_threadapi_pthread.c
+../../src/umock_lock_factory_pthread.c
+)
+endif()
+
+set(${theseTestsName}_c_files
+../../src/umock_c.c
```
Comment (mattdurak) [inline tests/umock_c_with_lock_factory_int/CMakeLists.txt:27]: any reason not to just use the lib?

File: tests/umock_lock_factory_pthread_ut/umock_lock_factory_pthread_ut.c [hunk 13]
```diff
@@ -0,0 +1,511 @@
...
+        mocked_calls = NULL;
+    }
+
+    mocked_call_count = 0;
+}
+
+static TEST_MUTEX_HANDLE test_mutex;
+static TEST_MUTEX_HANDLE global_mutex;
+
+BEGIN_TEST_SUITE(umock_lock_factory_windows_unittests)
```
Comment (mattdurak) [inline tests/umock_lock_factory_pthread_ut/umock_lock_factory_pthread_ut.c:233]: >_windows [](start = 35, length = 8)

pthread

**Review by mattdurak — APPROVED** [review]
:shipit:

File: devdoc/umock_lock_factory_windows.md (hunk 1 above)
Comment (dcristoloveanu) [reply devdoc/umock_lock_factory_windows.md:20]: Fixed.

---
In reply to: [530647559](https://github.com/Azure/umock-c/pull/177#discussion_r530647559) [](ancestors = 530647559)

File: devdoc/umock_lock_factory_pthread.md (hunk 2 above)
Comment (dcristoloveanu) [reply devdoc/umock_lock_factory_pthread.md:20]: Fixed.

---
In reply to: [530647904](https://github.com/Azure/umock-c/pull/177#discussion_r530647904) [](ancestors = 530647904)

File: doc/umock_c.md (hunk 3 above)
Comment (dcristoloveanu) [reply doc/umock_c.md:1339]: Added

---
In reply to: [530651263](https://github.com/Azure/umock-c/pull/177#discussion_r530651263) [](ancestors = 530651263)

File: src/umock_lock_factory_pthread.c (hunk 5 above)
Comment (dcristoloveanu) [reply src/umock_lock_factory_pthread.c:29]: Fixed.

---
In reply to: [530652478](https://github.com/Azure/umock-c/pull/177#discussion_r530652478) [](ancestors = 530652478)

File: src/umockcallrecorder.c (hunk 6 above)
Comment (dcristoloveanu) [reply src/umockcallrecorder.c:68]: Changed.

---
In reply to: [530653057](https://github.com/Azure/umock-c/pull/177#discussion_r530653057) [](ancestors = 530653057)

File: src/umockcallrecorder.c (hunk 7 above)
Comment (dcristoloveanu) [reply src/umockcallrecorder.c:409]: You are right

---
In reply to: [530653997](https://github.com/Azure/umock-c/pull/177#discussion_r530653997) [](ancestors = 530653997)

File: src/umockcallrecorder.c (hunk 8 above)
Comment (dcristoloveanu) [reply src/umockcallrecorder.c:489]: Fixed.

---
In reply to: [530654384](https://github.com/Azure/umock-c/pull/177#discussion_r530654384) [](ancestors = 530654384)

File: src/umockcallrecorder.c (hunk 9 above)
Comment (dcristoloveanu) [reply src/umockcallrecorder.c:611]: Changed, thank you!

---
In reply to: [530654785](https://github.com/Azure/umock-c/pull/177#discussion_r530654785) [](ancestors = 530654785)

File: tests/umockcallrecorder_ut/CMakeLists.txt (hunk 10 above)
Comment (dcristoloveanu) [reply tests/umockcallrecorder_ut/CMakeLists.txt:14]: Removed, was a mistake

---
In reply to: [530656573](https://github.com/Azure/umock-c/pull/177#discussion_r530656573) [](ancestors = 530656573)

File: tests/umock_c_with_lock_factory_int/minipal/umock_threadapi_pthread.c (hunk 11 above)
Comment (dcristoloveanu) [reply tests/umock_c_with_lock_factory_int/minipal/umock_threadapi_pthread.c:4]: [https://stackoverflow.com/questions/29201515/what-does-d-default-source-do](https://stackoverflow.com/questions/29201515/what-does-d-default-source-do)

---
In reply to: [530657003](https://github.com/Azure/umock-c/pull/177#discussion_r530657003) [](ancestors = 530657003)

File: tests/umock_c_with_lock_factory_int/CMakeLists.txt (hunk 12 above)
Comment (dcristoloveanu) [reply tests/umock_c_with_lock_factory_int/CMakeLists.txt:27]: There is another task by which I'll do that. I do not recall the exact reason it got coded like this in the first place.

---
In reply to: [530658693](https://github.com/Azure/umock-c/pull/177#discussion_r530658693) [](ancestors = 530658693)

File: tests/umock_lock_factory_pthread_ut/umock_lock_factory_pthread_ut.c (hunk 13 above)
Comment (dcristoloveanu) [reply tests/umock_lock_factory_pthread_ut/umock_lock_factory_pthread_ut.c:233]: Fixed.

---
In reply to: [530660377](https://github.com/Azure/umock-c/pull/177#discussion_r530660377) [](ancestors = 530660377)

**Pull request merged**